---
title: "Clinical Trait Scan"
author: "Brian Yandell"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
params:
  physiotrait: "percent_fat_17wk"
  genetrait: "Glb1"
  isoformid: "217583"
  chr: "9"
  width: 3
  lod_drop: 3
  snp36: TRUE
  datadir: "data"

---

```{r setup, include=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```
From: Tara Price <trprice@wisc.edu>
Date: Wednesday, December 13, 2023 at 4:02 PM
To: Brian Yandell <brian.yandell@wisc.edu>
Subject: Puzzling Cholesterol QTL
Hi Brian, 
Attaching a powerpoint with QTL scans and BLUPs for the cholesterol QTL on Chromosome 5. The first 5 slides are the QTL and their allele effects and how we found a candidate gene based on matching allele effects (old DO) and biological plausibility. Then, we ran the eQTL for Scarb1 using liver RNA from the new, diet DO study and the allele effects did not match the phenotype or old DO. 

The latter half of the slides are just pictures I generated from running the QTL scans/BLUPs when the two diets were separated. Essentially, there were slight differences in the SNP plots between the two diets, but the BLUPs looked very similar. I also added Scarb1 as a covariate to my scans, but that had very little effect on the cholesterol QTL. 

This slide deck focuses on the Chromosome 5 QTL, but I am also having difficulty identifying a plausible candidate on Chromosome 1 (despite Apoa2, a known cholesterol protein, being at that same locus). 

Any guidance/insight/suggestions you can provide will be GREATLY appreciated.
Please let me know if you need other information from me. I’m also happy to arrange a time and talk through my triage steps, if that’s useful.
```

```
From: MARK P KELLER <mark.keller@wisc.edu>
Date: Friday, December 8, 2023 at 3:52 PM
To: Brian Yandell <brian.yandell@wisc.edu>
Subject: percent_fat_17wk

Based on literature, the most plausible gene under the QTL is: Glb1.
 
There are only two isoforms for Glb1:
ENSMUST00000063042
ENSMUST00000217583
 
The one highlighted yellow has a cis-eQTL with a similar allele pattern as the fast mass QTL; the other does not.
```

# Data Setup

```{r}
library(dplyr)
library(tidyr)
library(ggplot2)
library(qtl2)
library(qtl2fst)
library(qtl2ggplot)
```

```{r}
chrname <- params$chr
physioname <- params$physiotrait
genename <- params$genetrait
isoformid <- params$isoformid
lod_drop <- params$lod_drop
chrwidth <- params$width
snp36 <- as.logical(params$snp36) # use 36 genotypes for SNPs
datadir <- as.character(params$datadir)
```

Clinical trait `r physioname` with covariate `r genename` (optional isoform `r isoformid`) on chr `r chrname`.
Window half-width of `r chrwidth` with SNP LOD drop of `r lod_drop` based on
`r ifelse(snp36, "36 allele pairs", "8 alleles")`.
Data is folder `datadir`.

# Dataset Setup

```{r}
datadir <- "data"

## 8 allele probabilities

fst_dir <- file.path(datadir, "fst_genoprob")
aprobs_chr <- subset(readRDS(file.path(fst_dir, "aprobs_fstindex.rds")), chr = chrname)

## Physical Map and Kinship

# data/gigamuga_map_test_v2.RDS
pmap <- readRDS(file.path(datadir, "pmap.rds"))

# data/K_1176_DietDO.rds
kinship_chr <- readRDS(file.path(datadir, "kinship.rds"))[chrname]

## Covariates

# data/Phenotype files/clinical data sets/Final in vivo DO data for analysis (raw and computed).xlsx
Physio <- readxl::read_excel(file.path(datadir, "Physio.xlsx"), sheet = "Data") |>
  tibble::column_to_rownames("mouse")
addcovar <- stats::model.matrix(
  ~ Gen_Lit + Sex * Diet.name,
  Physio |> dplyr::select(Gen_Lit, Sex, Diet.name))[,-1]
intcovar <- addcovar[,"Diet.nameHF", drop = FALSE]

physio_trait <- Physio[, grep(paste(physioname, collapse = "|"), colnames(Physio)), drop = FALSE]

## RNA Data

# Read gene data and combine gene name and ENSEMBL ID.
gene_trait <- readRDS("data/gene.rds")

gene_trait <- gene_trait[,
  grep(paste(paste0(genename, "_"), collapse = "|"), colnames(gene_trait)),
  drop = FALSE]
```

Isoform data

```{r}
isoform_trait <- readRDS("data/isoform.rds")
isoformname <- grep(paste(paste0(genename, "_"), collapse = "|"),
                          colnames(isoform_trait))
if(isoformid != "")
  isoformname <- isoformname[grep(isoformid, colnames(isoform_trait)[isoformname])]

isoform_trait <- isoform_trait[, isoformname, drop = FALSE]
```

# Trait Scan

```{r}
trait_addscan_chr <- qtl2::scan1(aprobs_chr, physio_trait, kinship_chr, addcovar,
                             cores = 0)

trait_intscan_chr <- qtl2::scan1(aprobs_chr, physio_trait, kinship_chr, addcovar,
                             intcovar = intcovar, cores = 0)

plot(trait_intscan_chr, pmap, col = "red")
plot(trait_addscan_chr, pmap, add = TRUE)
```

```{r}
(trait_peaks <- find_peaks(trait_intscan_chr, pmap, threshold = 9, drop = 2))
```

## Zoom in to small window

```{r}
plot(trait_intscan_chr, pmap, col = "red",
     xlim = trait_peaks$pos + c(-1,1) * chrwidth)
plot(trait_addscan_chr, pmap, add = TRUE,
     xlim = trait_peaks$pos + c(-1,1) * chrwidth)
abline(v = trait_peaks$pos, col = "gray", lwd = 2)
```

# Strain Coefficients

```{r}
trait_nocoef <- scan1coef(aprobs_chr, physio_trait, kinship_chr)
plot_coef(trait_nocoef, pmap[chrname], columns = 1:8)
```

```{r}
trait_addcoef <- scan1coef(aprobs_chr, physio_trait, kinship_chr, addcovar)
plot_coef(trait_addcoef, pmap[chrname], columns = 1:8,
          xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "grey", lwd = 2)
title(paste(physioname, "on chr", chrname))
```

The following plot is flawed as it only presents part of the solution.

```{r}
trait_intcoef <- scan1coef(aprobs_chr, physio_trait, kinship_chr, addcovar,
                           intcovar = intcovar)

plot_coef(trait_intcoef, pmap[chrname], columns = 1:8,
          xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "grey", lwd = 2)
title(paste(physioname, "on chr", chrname, "additive part of diet interaction"))
```

Following does separate coefficients by diet. 
It is complicated by fact that `kinship` calculation does not include one mouse.

```{r}
wh_HF <- (addcovar[,"Diet.nameHF"] == 1)
m <- match(rownames(addcovar), rownames(kinship_chr[[1]]))
#rownames(addcovar)[is.na(m)]
id_HF <- rownames(addcovar)[wh_HF & !is.na(m)]
id_HC <- rownames(addcovar)[!wh_HF & !is.na(m)]

trait_coef_HF <- scan1coef(
  subset(aprobs_chr, ind = id_HF),
  physio_trait[id_HF,, drop = FALSE],
  kinship_chr, addcovar, intcovar = intcovar)

plot_coef(trait_coef_HF, pmap[chrname], columns = 1:8,
          xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "grey", lwd = 2)
title(paste("HF Diet", physioname, "on chr", chrname))

trait_coef_HC <- scan1coef(
  subset(aprobs_chr, ind = id_HC),
  physio_trait[id_HC,, drop = FALSE],
  kinship_chr, addcovar, intcovar = intcovar)

plot_coef(trait_coef_HC, pmap[chrname], columns = 1:8,
          xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "grey", lwd = 2)
title(paste("HC Diet", physioname, "on chr", chrname))
```

# SNP Associations

SNP genotype probabilities are calculated as 2-level additive (using `aprobs_chr` from the 8-allele genotype probablities) or 3-level full (using `probs_chr` from the 36-allele-pair genotype probabilities).
The latter are used with `snp36` is `TRUE` (default) and the 36-pair
probabilities are available.

```{r}
library(ggplot2)
library(qtl2ggplot)
```

```{r}
chrpos <- trait_peaks$pos
```

```{r}
snpdb_file <- "data/cc_variants.sqlite"
query_variant <- qtl2::create_variant_query_func(snpdb_file)
snpinfo <- query_variant(chrname,
                         chrpos - chrwidth,
                         chrpos + chrwidth)
```

Need to change names of alleles

```{r}
names(snpinfo)[8:15] <- names(qtl2::CCcolors)
names(snpinfo)[1] <- "snp_id"
```

Perform SNP association mapping.
It is possible to use `qtl2::scan1snps` instead, which bundles these three routines,
but we want to have the SNP probabilities for later use.

```{r}
snpinfo <- qtl2::index_snps(pmap, snpinfo)
```

```{r}
if(snp36 & file.exists(file.path(datadir, "fst_genoprob36"))) {
  snppr <- qtl2::genoprob_to_snpprob(
    subset(
      readRDS(file.path(datadir, "fst_genoprob36", "probs_fstindex.rds")),
      chr = chrname),
    snpinfo)
} else {
  snppr <- qtl2::genoprob_to_snpprob(aprobs_chr, snpinfo)
}
```

```{r}
scan_snppr <- qtl2::scan1(snppr, physio_trait, kinship_chr, addcovar,
                          intcovar = intcovar)
```

```{r}
autoplot(scan_snppr, snpinfo, drop_hilit = lod_drop)
```

```{r}
autoplot(scan_snppr, snpinfo, patterns="hilit", drop_hilit=lod_drop, cex=2)
```

```{r}
autoplot(scan_snppr, snpinfo, patterns="all", drop_hilit=lod_drop, cex=2)
```


```{r}
(snp_peak <- qtl2::find_peaks(scan_snppr, snpinfo))
```

Find top snps for distinct SDPs within `r `lod_drop` of peak.

```{r}
(top_snps_tbl <-
   qtl2pattern::top_snps_pattern(scan_snppr, snpinfo,
                                 drop = lod_drop,
                                 show_all_snps = FALSE) |>
  select(snp_id, pos, sdp, index, lod) |>
  distinct(sdp, .keep_all = TRUE) |>
  arrange(desc(lod)))
```

# Use best SNPs as covariate

```{r}
snp_id <- top_snps_tbl$snp_id
snp_df <- dim(snppr[[1]])[2] - 1
snp_geno <- matrix(NA, nrow(addcovar), length(snp_id) * snp_df)
rownames(snp_geno) <- rownames(addcovar)
if(snp_df == 1) {
  colnames(snp_geno) <- snp_id
} else {
  colnames(snp_geno) <- paste(rep(snp_id, each = 2), seq(snp_df), sep = "_")
}
snp_geno[dimnames(snppr[[1]])[[1]],] <- snppr[[1]][, -(1 + snp_df), snp_id]
```

```{r}
addcovar_geno <- cbind(addcovar, snp_geno)
addcovar_geno1 <- cbind(addcovar, snp_geno[,seq(snp_df), drop = FALSE])
```

```{r}
trait_genoscan_chr <- qtl2::scan1(aprobs_chr, physio_trait,
                                  kinship_chr, addcovar_geno,
                                  intcovar = intcovar, cores = 10)
trait_geno1scan_chr <- qtl2::scan1(aprobs_chr, physio_trait,
                                  kinship_chr, addcovar_geno1,
                                  intcovar = intcovar, cores = 10)
```

```{r}
bind_rows(
  add = qtl2::find_peaks(trait_addscan_chr, pmap),
  int = qtl2::find_peaks(trait_intscan_chr, pmap),
  geno1 = qtl2::find_peaks(trait_geno1scan_chr, pmap),
  geno = qtl2::find_peaks(trait_genoscan_chr, pmap),
  .id = "scan")
```

```{r}
plot(trait_intscan_chr, pmap, col = "red",
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_genoscan_chr, pmap, col = "blue", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_geno1scan_chr, pmap, col = "purple", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_addscan_chr, pmap, col = "green", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "gray", lwd = 2)
abline(v = snp_peak$pos, col = "blue", lwd = 2)
```

# Add gene and isoform as covariate

```{r}
dat <- as.data.frame(cbind(physio_trait, gene_trait, isoform_trait))
names(dat) <- c(physioname, genename, isoformid)
ggplot(dat) +
  aes(.data[[physioname]], .data[[genename]]) +
  geom_point() +
  geom_smooth(se = FALSE)
```

```{r}
ggplot(dat) +
  aes(.data[[physioname]], .data[[isoformid]]) +
  geom_point() +
  geom_smooth(se = FALSE)
```

```{r}
ggplot(dat) +
  aes(.data[[genename]], .data[[isoformid]]) +
  geom_point() +
  geom_smooth(se = FALSE)
```

```{r}
addcovar_gene <- cbind(addcovar, gene_trait)
addcovar_isoform <- cbind(addcovar, isoform_trait)
addcovar_gene_isoform <- cbind(addcovar, gene_trait, isoform_trait)
```

```{r}
trait_genescan_chr <- qtl2::scan1(aprobs_chr, physio_trait,
                                  kinship_chr, addcovar_gene,
                                  intcovar = intcovar, cores = 10)
trait_isoformscan_chr <- qtl2::scan1(aprobs_chr, physio_trait,
                                  kinship_chr, addcovar_isoform,
                                  intcovar = intcovar, cores = 10)
trait_gene_isoscan_chr <- qtl2::scan1(aprobs_chr, physio_trait,
                                  kinship_chr, addcovar_gene_isoform,
                                  intcovar = intcovar, cores = 10)
```

```{r}
bind_rows(
  add = qtl2::find_peaks(trait_addscan_chr, pmap),
  int = qtl2::find_peaks(trait_intscan_chr, pmap),
  gene = qtl2::find_peaks(trait_genescan_chr, pmap),
  isoform = qtl2::find_peaks(trait_isoformscan_chr, pmap),
  gene_iso = qtl2::find_peaks(trait_gene_isoscan_chr, pmap),
  .id = "scan")
```

```{r}
plot(trait_intscan_chr, pmap, col = "red",
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_genescan_chr, pmap, col = "blue", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_isoformscan_chr, pmap, col = "purple", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_gene_isoscan_chr, pmap, col = "black", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
plot(trait_addscan_chr, pmap, col = "green", add = TRUE,
     xlim = trait_peaks$pos + c(-1,1))
abline(v = trait_peaks$pos, col = "gray", lwd = 2)
```

# Mediation

```{r}

```

```{r}
knitr::knit_exit()
```

## qtl2pattern

scan1pattern does not pass on incovar yet.
It also does not seem to work properly.
It should use the SDP in snpinfo identified to index snp to impute SNP probabilities
for pattern. It may be set up to do this with pmap but have some sort of disconnect in code.

```{r}
top_snps_tbl <- qtl2pattern::top_snps_pattern(scan_snppr, snpinfo)
```

```{r}
(patterns <- summary(top_snps_tbl))
```

```{r}
head(summary(top_snps_tbl, "best"))
```

```{r}
scan_pat <- qtl2pattern::scan1pattern(snppr, physio_trait, kinship_chr, addcovar,
                                      map = pmap, patterns = patterns)
```


```{r}
autoplot(scan_pat$scan, snpinfo, patterns = "hilit")
```

```{r}
summary(scan_pat$scan, pmap, chr = chrname)
```

```{r}
qtl2::find_peaks()
```

```{r}
tmp <- snpinfo[!duplicated(snpinfo$snp_id), "pos"]
names(tmp) <- unique(snpinfo$snp_id)
tmp <- list("19" = tmp)
```

```{r}
qtl2pattern::ggplot_scan1pattern(scan_pat, tmp)
```
